pkl-core/src/main/java/org/pkl/core/externalReader/ExternalReaderProcessImpl.java

```diff
@@ -0,0 +1,222 @@
+  private final Map<String, Future<@Nullable ModuleReaderSpec>> initializeModuleReaderResponses =
+      new ConcurrentHashMap<>();
+  private final Map<String, Future<@Nullable ResourceReaderSpec>>
+      initializeResourceReaderResponses = new ConcurrentHashMap<>();
+
+  private @GuardedBy("this") boolean closed = false;
+
+  @LateInit
+  @GuardedBy("this")
+  private Process process;
+
+  @LateInit
+  @GuardedBy("this")
+  private MessageTransport transport;
+
+  private void log(String msg) {
+    if (logPrefix != null) {
+      System.err.println(logPrefix + msg);
+    }
+  }
+
+  public ExternalReaderProcessImpl(ExternalReader spec) {
+    this.spec = spec;
+    logPrefix =
+        Objects.equals(System.getenv("PKL_DEBUG"), "1")
+            ? "[pkl-core][external-process][" + spec.executable() + "] "
+            : null;
+  }
+
+  @Override
+  public synchronized MessageTransport getTransport() throws ExternalReaderProcessException {
+    if (closed) {
+      throw new ExternalReaderProcessException("ExternalProcessImpl has already been closed");
+    }
+    if (process != null) {
+      if (!process.isAlive()) {
+        throw new ExternalReaderProcessException("ExternalProcessImpl process is no longer alive");
+      }
+
+      return transport;
+    }
+
+    // This relies on Java/OS behavior around PATH resolution, absolute/relative paths, etc.
+    var command = new ArrayList<String>();
+    command.add(spec.executable());
+    command.addAll(spec.arguments());
+
+    var builder = new ProcessBuilder(command);
+    builder.redirectError(Redirect.INHERIT); // inherit stderr from this pkl process
+    try {
+      process = builder.start();
+    } catch (IOException e) {
+      throw new ExternalReaderProcessException(e);
+    }
+    transport =
+        MessageTransports.stream(
+            new ExternalReaderMessagePackDecoder(process.getInputStream()),
+            new ExternalReaderMessagePackEncoder(process.getOutputStream()),
+            this::log);
+
+    var rxThread = new Thread(this::runTransport, "ExternalProcessImpl rxThread for " + spec);
+    rxThread.setDaemon(true);
+    rxThread.start();
+
+    return transport;
+  }
+
+  /**
+   * Runs the underlying message transport so it can receive responses from the child process.
+   *
+   * <p>Blocks until the underlying transport is closed.
+   */
+  private void runTransport() {
+    try {
+      transport.start(
+          (msg) -> {
+            throw new ProtocolException("Unexpected incoming one-way message: " + msg);
+          },
+          (msg) -> {
+            throw new ProtocolException("Unexpected incoming request message: " + msg);
+          });
+    } catch (ProtocolException | IOException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  @Override
+  public synchronized void close() {
+    closed = true;
+    if (process == null || !process.isAlive()) {
+      return;
+    }
+
+    try {
+      getTransport().send(new CloseExternalProcess());
```

Comment:
Should be fine to close the transport at this point too? This will remove any response handlers.

```suggestion
      getTransport().send(new CloseExternalProcess());
      getTransport().close();
```